toolchain/semantics/semantics_handle_index.cpp

```diff
@@ -27,21 +26,40 @@ auto SemanticsHandleIndexExpression(SemanticsContext& context,
       context.semantics_ir().GetTypeAllowBuiltinTypes(name_node.type_id());
   auto name_type_node = context.semantics_ir().GetNode(name_type_id);
 
-  if (name_type_node.kind() == SemanticsNodeKind::TupleType &&
-      index_node.kind() == SemanticsNodeKind::IntegerLiteral) {
+  if (name_type_node.kind() == SemanticsNodeKind::ArrayType) {
+    auto [bound_id, type_id] = name_type_node.GetAsArrayType();
+    if (index_node.kind() == SemanticsNodeKind::IntegerLiteral) {
+      const auto& index_val = context.semantics_ir().GetIntegerLiteral(
+          index_node.GetAsIntegerLiteral());
+      if (index_val.uge(context.semantics_ir().GetArrayBoundValue(bound_id))) {
+        context.DiagnoseOutOfBounds(parse_node, index_val, name_node);
+      } else {
+        context.AddNodeAndPush(
+            parse_node, SemanticsNode::ArrayIndex::Make(
+                            parse_node, type_id, name_node_id, index_node_id));
+        return true;
+      }
+    } else if (SemanticsBuiltinKind::FromInt(
+                   context.semantics_ir()
+                       .GetTypeAllowBuiltinTypes(index_node.type_id())
+                       .index) ==
+               Internal::SemanticsBuiltinKindRawEnum::IntegerType) {
+      context.AddNodeAndPush(
+          parse_node, SemanticsNode::ArrayIndex::Make(
+                          parse_node, type_id, name_node_id, index_node_id));
+      return true;
+    } else if (name_type_id != SemanticsNodeId::BuiltinError) {
+      context.DiagnoseUndeterministicType(parse_node);
+    }
```

Comment:
For arrays, we generally want to allow any index that can be converted to an integer. So this should probably be calling ImplicitAsRequired to try converting index_node to IntegerType, versus the current requirement that index_node is already an IntegerType.